Clean up and more liberally free holding cell HTLCs (without re-entrancy)

TheBlueMatt opened the pull request:
This is #756 reworked a decent chunk. In #756 the reentrancy story around `channel_monitor_updated` scared me significantly, but for some reason I apparently never bothered to really think deeply about the options. Here we move the holding cell free to `get_and_clear_pending_msg_events`, which is a *super* obvious location for it - it should be called right after `channel_monitor_updated` in most cases, but shouldn't have the same reentrancy issues because it already can generate a monitor update. It requires some macro rework, but its not too bad.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -2692,13 +2692,11 @@ impl<Signer: Sign> Channel<Signer> {
 		if self.channel_state < ChannelState::FundingSent as u32 {
 			self.channel_state = ChannelState::ShutdownComplete as u32;
-			return outbound_drops;
+			return;
```
If this is gonna be a pretty first-class use case that we expect everyone to use, should we also consider not dropping e.g. `RemoteAnnounced` HTLCs below?

TheBlueMatt replied:
We aren't actually rejecting those HTLCs, only dropping them. If we reconnect (and the peer hasn't dropped them), then they'll re-send them.

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -2741,23 +2739,8 @@ impl<Signer: Sign> Channel<Signer> {
 		self.channel_state |= ChannelState::PeerDisconnected as u32;
-		log_debug!(logger, "Peer disconnection resulted in {} remote-announced HTLC drops and {} waiting-to-locally-announced HTLC drops on channel {}", outbound_drops.len(), inbound_drop_count, log_bytes!(self.channel_id()));
-		outbound_drops
+		log_debug!(logger, "Peer disconnection resulted in {} waiting-to-locally-announced HTLC drops on channel {}", inbound_drop_count, log_bytes!(self.channel_id()));
```
I think the log is supposed to be: `Peer disconnection resulted in {} remote-announced HTLC drops on channel {}` instead

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3045,77 +3076,34 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-							order = RAACommitmentOrder::RevokeAndACKFirst;
-						}
-						return_monitor_err!(self, e, channel_state, chan, order, revoke_and_ack.is_some(), commitment_update.is_some());
-						//TODO: Resend the funding_locked if needed once we get the monitor running again
```
Hmm, thats a good question!  I don't *think* it matters - this PR switched the ordering of message resending to resend the funding_locked before handling the monitor update, meaning we'll never need to resend because we always send (instead of maybe not sending if the monitor update fails).

The "don't send a funding_locked if we're paused pending monitor update" thing isn't really a security thing (we just expose a `next_per_commitment_point` which doesn't allow our counterparty to....do anything), its just kinda nice to not tell our counterparty we're ready to accept HTLCs when we're definitely not. That said, `Channel::channel_reestablish` doesn't really handle it all that carefully anyway, and we may end up re-sending a funding_locked on reconnect when we are still pending a monitor update failure.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -786,20 +786,60 @@ macro_rules! maybe_break_monitor_err {
+					$channel_entry.get().get_funding_txo().unwrap(), forwards));
+			}
+
+			let chanmon_update: Option<ChannelMonitorUpdate> = $chanmon_update; // Force type-checking to resolve
+			if chanmon_update.is_some() {
+				assert!($commitment_update.is_some());
+				assert!($funding_locked.is_none());
+			}
```
Ok, I spent way too long trying to write a test to break this and I'm pretty confident in it :)

valentinewallace replied:
Oh, I was just asking for a comment^ lol

TheBlueMatt replied:
Well, hey, I started writing the comment and then wanted to be sure I was right :)

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -786,20 +786,60 @@ macro_rules! maybe_break_monitor_err {
-	($self: expr, $channel_lock: expr, $channel_state: expr, $channel_entry: expr,
+	($self: ident, $channel_lock: expr, $channel_state: expr, $channel_entry: expr,
 	 $raa: expr, $commitment_update: expr, $order: expr,
+	 $pending_forwards: expr, $broadcast_safe: expr, $funding_locked: expr) => { {
```
Do we have a standard for more-parameters-indentation? The one-space indentation here it to line up the parameters, but I can do the double-tab that I think we do in some other places if you prefer.

valentinewallace replied:
Oops, I meant "nit: inconsistent line lengths"

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2336,87 +2459,24 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+			let (raa, commitment_update, order, pending_forwards, pending_failures, needs_broadcast_safe, funding_locked) = channel.get_mut().monitor_updating_restored(&self.logger);
+			(pending_failures, handle_chan_restoration_locked!(self, channel_lock, channel_state, channel, raa, commitment_update, order, pending_forwards, needs_broadcast_safe, funding_locked))
+		};
+		post_handle_chan_restoration!(self, chan_restoration_res);
```
The `handle_chan_restoration_locked` macro does this for us.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -666,22 +666,43 @@ macro_rules! handle_error {
 				}
-				break Err(MsgHandleErrInternal::from_finish_shutdown(msg, channel_id, chan.force_shutdown(true), $self.get_channel_update(&chan).ok()))
-			},
-			Err(ChannelError::CloseDelayBroadcast(_)) => { panic!("Wait is only generated on receipt of channel_reestablish, which is handled by try_chan_entry, we don't bother to support it here"); }
```
Right, in combining them we call the same version either way, so we have to have the code for it there. I dont think there's any harm in it, its dead code for the places that previously had an assertion.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3211,6 +3217,52 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 	}
 
+	/// Process pending events from the `chain::Watch`.
+	fn check_free_holding_cells(&self) {
```
Hmm, its a reference to each channel having its own cell, does it matter?

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -2350,7 +2357,9 @@ impl<Signer: Sign> Channel<Signer> {
 						match self.get_update_fulfill_htlc(htlc_id, *payment_preimage, logger) {
 							Ok((update_fulfill_msg_option, additional_monitor_update_opt)) => {
-								update_fulfill_htlcs.push(update_fulfill_msg_option.unwrap());
+								if let Some(update_fulfill_msg) = update_fulfill_msg_option { //WTF XXX
```
lol, that shouldn't have slipped in there, the hunk is bogus (I personally use "XXX" for "don't ever commit this/merge this, so its just my fault for letting it go in the PR).

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3211,6 +3217,52 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		}
 	}
 
+	/// Process pending events from the `chain::Watch`.
```
Not even close :)

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -2300,6 +2300,13 @@ impl<Signer: Sign> Channel<Signer> {
+	/// Public version of the below, checking relevant preconditions first
+	pub fn maybe_free_holding_cell_htlcs<L: Deref>(&mut self, logger: &L) -> Result<(Option<(msgs::CommitmentUpdate, ChannelMonitorUpdate)>, Vec<(HTLCSource, PaymentHash)>), ChannelError> where L::Target: Logger {
+		if self.channel_state >= ChannelState::ChannelFunded as u32 &&
+		   (self.channel_state & (ChannelState::AwaitingRemoteRevoke as u32 | ChannelState::PeerDisconnected as u32 | ChannelState::MonitorUpdateFailed as u32)) == 0 {
```
Hmmm, I don't think so - on a graceful shutdown start we drop HTLC adds and wait until later for the HTLC fulfill/fails, but we still want to send those - ultimately we're waiting on those to finish so that we can do a graceful shutdown. Once we're "really" shut down (either forced or done negotiating the graceful shutdown), we drop the Channel object so we can't get here anyway.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -765,11 +764,18 @@ macro_rules! handle_monitor_err {
 		}
-	}
+	};
+	($self: ident, $err: expr, $channel_state: expr, $entry: expr, $action_type: path, $resend_raa: expr, $resend_commitment: expr, $failed_forwards: expr, $failed_fails: expr) => { {
+		let (res, drop) = handle_monitor_err!($self, $err, $channel_state.short_to_id, $entry.get_mut(), $action_type, $resend_raa, $resend_commitment, $failed_forwards, $failed_fails, $entry.key());
+		if drop {
+			$entry.remove_entry();
+		}
+		res
+	} };
```
We could check for it at the callsite, but that ends up with flow graph information in two completely separate places, whereas at least when its here you can read the flowgraph easier (or, at least, I can). If you feel strongly we can move it (or we could do a boolean parameter, though that's also confusing IMO).

valentinewallace replied:
Eh yeah, I think it's OK as is. I mainly just find `commitment_signed_dance` confusing but I think this is acceptable.

TheBlueMatt replied:
Yea, `commitment_signed_dance` could *definitely* use a cleanup.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -786,20 +786,60 @@ macro_rules! maybe_break_monitor_err {
 }
 
 macro_rules! handle_chan_restoration_locked {
```
Hmm, yea, it does refer to the `channel_state` lock, I guess I'm just afraid of getting the locking wrong in the future because we've had awkward locking incorrectness bugs in the past. If you feel strongly I can drop it.

valentinewallace replied:
I don't feel strongly, resolving!

valentinewallace reviewed (commented):
Nothing major. This looks pretty reasonable to me!

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -2300,6 +2300,13 @@ impl<Signer: Sign> Channel<Signer> {
 
+	/// Public version of the below, checking relevant preconditions first
+	pub fn maybe_free_holding_cell_htlcs<L: Deref>(&mut self, logger: &L) -> Result<(Option<(msgs::CommitmentUpdate, ChannelMonitorUpdate)>, Vec<(HTLCSource, PaymentHash)>), ChannelError> where L::Target: Logger {
+		if self.channel_state >= ChannelState::ChannelFunded as u32 &&
+		   (self.channel_state & (ChannelState::AwaitingRemoteRevoke as u32 | ChannelState::PeerDisconnected as u32 | ChannelState::MonitorUpdateFailed as u32)) == 0 {
```
Hmm, we already check all the preconditions at the callsites, I suppose we could move to just calling `maybe_...`, but that seems more brittle, given we already check them.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3221,6 +3227,52 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+							if let Err(e) = self.chain_monitor.update_channel(chan.get_funding_txo().unwrap(), monitor_update) {
+								let (res, drop) = handle_monitor_err!(self, e, short_to_id, chan, RAACommitmentOrder::CommitmentFirst, false, true, Vec::new(), Vec::new(), channel_id);
+								handle_errors.push((chan.get_counterparty_node_id(), res));
+								if drop { return false; }
```
I renamed `drop` to `close_channel` is that more clear?

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3221,6 +3227,52 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+		}
+
+		for (counterparty_node_id, err) in handle_errors.drain(..) {
+			let _ = handle_error!(self, err, counterparty_node_id);
```
Hmm, right, that's awkward, but I think also globally true - we always generate a `ChannelForceClosed` update for the monitor, even if the user can't persist it.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -3029,20 +3004,14 @@ impl<Signer: Sign> Channel<Signer> {
-						// disconnect, we may be handed some HTLCs to fail backwards here.
-						assert!(htlcs_to_fail.is_empty());
-						return Ok((resend_funding_locked, required_revoke, Some(commitment_update), Some(monitor_update), self.resend_order.clone(), shutdown_msg));
+						return Ok((resend_funding_locked, required_revoke, Some(commitment_update), Some(monitor_update), self.resend_order.clone(), htlcs_to_fail, shutdown_msg));
```
Hmm, possibly. The holding cell is only really intended for things that are just pending some network latency - you should update a channel monitor or receive a revoke_and_ack from your peer within some hundreds of milliseconds, at max - not for things that are going to take a while - ie its *not* (really) for when your counterparty is offline.

valentinewallace reviewed (commented):
I'm ACK af14d44a0cb8c62b081cfda8f30f42d82fb171dd modulo fixing 5613be8146fde46460c0f48bd6a517fa6160128b's commit message

TheBlueMatt commented:
Addressed Val's comments, full diff is:
```
$ git diff-tree -U1 af14d44 386cc60
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index c40e7f1c4..a2bcede85 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -2309,2 +2309,3 @@ impl<Signer: Sign> Channel<Signer> {
 	}
+
 	/// Used to fulfill holding_cell_htlcs when we get a remote ack (or implicitly get it by them
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 73ba9d9bd..b538655bd 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -731,3 +731,3 @@ macro_rules! handle_monitor_err {
 			ChannelMonitorUpdateErr::PermanentFailure => {
-				log_error!($self.logger, "Closing channel {} due to monitor update PermanentFailure", log_bytes!($chan_id[..]));
+				log_error!($self.logger, "Closing channel {} due to monitor update ChannelMonitorUpdateErr::PermanentFailure", log_bytes!($chan_id[..]));
 				if let Some(short_id) = $chan.get_short_channel_id() {

```

TheBlueMatt commented:
Rebased on latest upstream.

TheBlueMatt commented:
Rebased on upstream, with pretty trivial changes.

TheBlueMatt commented:
Looks like this is slipping to 0.0.15. It would be good to land some of these fixes as they are potential bugs, even if very rare.

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -815,19 +815,73 @@ macro_rules! maybe_break_monitor_err {
 			macro_rules! handle_cs { () => {
+				if let Some(monitor_update) = chanmon_update {
+					assert!($order == RAACommitmentOrder::RevokeAndACKFirst);
+					assert!(funding_broadcastable.is_none());
```
"According to the spec, if the funding transaction has not been broadcast, the funding node should forget about the channel (and thus not participate in a proposed reestablishment)"?

TheBlueMatt replied:
No, we only forget if we haven't received a funding_signed, not just not yet broadcast. I've added a comment, though.

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -815,19 +815,73 @@ macro_rules! maybe_break_monitor_err {
+			let funding_broadcastable: Option<Transaction> = $funding_broadcastable; // Force type-checking to resolve
 			macro_rules! handle_cs { () => {
+				if let Some(monitor_update) = chanmon_update {
+					assert!($order == RAACommitmentOrder::RevokeAndACKFirst);
```
I can't make my mind about correctness of this statement. It implies the path in `Channel::channel_reestablish` L3023, when we return htlc failures to route backward included in a new commitment transaction ? It doesn't necessary mean peer send us a commitment_signed for which we should send a RAA /

TheBlueMatt replied:
Yea, I can't convince myself its safe anymore, I walked this bit back.

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -695,22 +695,43 @@ macro_rules! handle_error {
-				if let Some(short_id) = chan.get_short_channel_id() {
-					$channel_state.short_to_id.remove(&short_id);
+			Err(e) => {
+				let (drop, res) = convert_chan_err!($self, e, $channel_state.short_to_id, $entry.get_mut(), $entry.key());
```
Note, you're loosing the panic! when `CloseDelayBroadcast` shouldn't show up in path where `break_chan_entry` is called. Though not sure if it's worthy to enrich the new `convert_chan_err` with a new boolean flag to panic when it's called from here... Likely more error-prone.

TheBlueMatt replied:
True, but we now handle it, so I'm not sure it matters?

ariard replied:
Yeah one day we'll make those errors macro better :p

ariard reviewed (commented):
Reviewed until e33012f included so far

TheBlueMatt commented on fuzz/src/chanmon_consistency.rs:
```diff
@@ -660,7 +659,6 @@ pub fn do_test<Out: test_logger::Output>(data: &[u8], out: Out) {
 							events::MessageSendEvent::SendChannelReestablish { .. } => {},
 							events::MessageSendEvent::SendFundingLocked { .. } => {},
 							events::MessageSendEvent::PaymentFailureNetworkUpdate { .. } => {},
-							events::MessageSendEvent::HandleError { action: ErrorAction::IgnoreError, .. } => {},
```
Added relevant panics to the `[fuzz] Fail in chanmon_consistency on IgnoreError error events` commit.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -853,30 +907,21 @@ macro_rules! handle_chan_restoration_locked {
 				},
 			}
-			if let Some(tx) = $funding_broadcastable {
+			if let Some(tx) = funding_broadcastable {
```
Hmm, that's true. I'm not sure its worth worrying about - the case where we do an async monitor update after a peer sent us a funding_signed but then the peer disconnects before we finish updating the monitor should be rare enough that I'm not sure its worth adding code to explicitly handle it.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -815,19 +815,73 @@ macro_rules! maybe_break_monitor_err {
 
+			let chanmon_update: Option<ChannelMonitorUpdate> = $chanmon_update; // Force type-checking to resolve
+			if chanmon_update.is_some() {
+				// On reconnect (or monitor restoration), we, by definition, only resend a
```
Hmm? Yes it does - it returns one if `Channel::monitor_pending_commitment_signed` is set.

TheBlueMatt replied:
Hmm, you're right. I removed the reference to monitor restoration, but I'm not sure your comment communicates enough information to later re-analyze the situation going on.

ariard replied:
Hmmmm "In case of monitor restoration, funding_locked will never be generated as this event isn't altering monitor state"?

TheBlueMatt commented:
Rebased on latest upstream, relevant changes responding to comments are in fixup commits.

ariard reviewed (commented):
Left few more comments, top commits sounds good to me, but want to do a second parse with fixed comments.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3413,6 +3419,57 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+				match chan.maybe_free_holding_cell_htlcs(&self.logger) {
+					Ok((None, ref htlcs)) if htlcs.is_empty() => true,
+					Ok((commitment_opt, failed_htlcs)) => {
+						forwarding_failed_htlcs.push((failed_htlcs, *channel_id));
```
Yea, that's confusing, the name was "htlcs which failed during forwarding" but that wasn't clear.

TheBlueMatt commented:
I believe I've addressed all the comments.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -913,23 +913,30 @@ macro_rules! handle_chan_restoration_locked {
-					assert!($commitment_update.is_some());
-					if let Err(e) = $self.chain_monitor.update_channel($channel_entry.get().get_funding_txo().unwrap(), monitor_update) {
-						break handle_monitor_err!($self, e, $channel_state, $channel_entry, RAACommitmentOrder::CommitmentFirst, false, true);
+			if let Some(monitor_update) = chanmon_update {
```
Sorry I missed it, what was the reasoning for adding this commit: e6b1635f07d159da565d1bc5baecfbe165c8c26b

TheBlueMatt replied:
See discussion at https://github.com/rust-bitcoin/rust-lightning/pull/851#discussion_r632094939

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -1379,13 +1379,15 @@ impl<Signer: Sign> Channel<Signer> {
 			return Ok(None);
 		}
 
+		log_trace!(logger, "Failure HTLC ID {} back with a update_fail_htlc message", htlc_id_arg);
```
Can't quite parse this log message... s/Failure/Failing?

TheBlueMatt replied:
Oops, yes, should be "Failing".

valentinewallace reviewed (commented):
From what I can see, nothing toooo crazy has changed since my last review. So this looks pretty reasonable! Might glance once more after CI is fixed.

TheBlueMatt commented:
Rebased and fixed, dropped one commit that added persistence notifiers which is duplicative with #920.

valentinewallace reviewed (approved):
LGTM :star2:

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -874,16 +874,74 @@ macro_rules! maybe_break_monitor_err {
-		{
-			if !$pending_forwards.is_empty() {
+		let chanmon_update: Option<ChannelMonitorUpdate> = $chanmon_update; // Force type-checking to resolve
+		let chanmon_update_is_some = chanmon_update.is_some();
```
Is this variable needed?

TheBlueMatt replied:
Yes, inside the loop we move out of chanmon_update with an `if let Some(...) = `.

jkczyz replied:
Ah, interesting use of `loop`. Maybe name it `chanmon_update_is_none` instead of negating later. Or name it `has_chanmon_update`.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -754,22 +754,43 @@ macro_rules! handle_error {
 
+macro_rules! convert_chan_err {
+	($self: ident, $err: expr, $short_to_id: expr, $channel: expr, $channel_id: expr) => {
+		match $err {
```
For readability, it would be helpful to use a variable here to know what the bool is meant to represent. Or add some documentation on the what the macro evaluates to.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -2701,13 +2701,11 @@ impl<Signer: Sign> Channel<Signer> {
-	pub fn remove_uncommitted_htlcs_and_mark_paused<L: Deref>(&mut self, logger: &L) -> Vec<(HTLCSource, PaymentHash)> where L::Target: Logger {
-		let mut outbound_drops = Vec::new();
-
+	pub fn remove_uncommitted_htlcs_and_mark_paused<L: Deref>(&mut self, logger: &L)  where L::Target: Logger {
```
Update docs with regard to return value. Move to `channel_reestablish`?

jkczyz reviewed (commented):
Nice clean-up! I need to take a closer look at the new test, but otherwise looks good to go.

TheBlueMatt commented:
Rebased with no changes aside from two new fixup commits to respond to Jeff's comments.

jkczyz reviewed (approved):
Looks good to me, but I'd be more comfortable if @ariard gave a look before merging.

TheBlueMatt commented:
Squashed and renamed one variable as @jkczyz suggested:
```
$ git diff-tree -U1 7fde8910 34fcd99f
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 14d9c528..9f9820c1 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -910,3 +910,3 @@ macro_rules! handle_chan_restoration_locked {
 		let chanmon_update: Option<ChannelMonitorUpdate> = $chanmon_update; // Force type-checking to resolve
-		let chanmon_update_is_some = chanmon_update.is_some();
+		let chanmon_update_is_none = chanmon_update.is_none();
 		let res = loop {
@@ -1006,3 +1006,3 @@ macro_rules! handle_chan_restoration_locked {
 
-		if !chanmon_update_is_some {
+		if chanmon_update_is_none {
 			// If there was no ChannelMonitorUpdate, we should never generate an Err in the res loop
```

TheBlueMatt commented:
Gonna go ahead and take this since its been a long time. Post-merge review from @ariard would also be welcome but I'd rather not wait on it if he's busy.
